### DISCUSSION TITLE
Can you recommend how to design a reusable widget with Riverpod as a reusable state?

### DISCUSSION THREAD

**Discussion by mgolokhov** [rrousselGit/riverpod#1052] [Q&A]
Hello,
Thank you for the great library :)

Can you recommend how to design a reusable widget with Riverpod as a reusable state?

As an example, could be a Table widget with

base actions:
- load
- refresh
- sortBy

base states:
- loading
- refreshing
- error
- errorWithData
- data

The first try was to use StateNotifier with states baked by Freezed package.  

How can I give through Table constructor base provider and then `ref.watch(baseTableProvider)` base states and trigger actions by `ref.read(baseTableProvider)`?

**Comment by rrousselGit** [comment]
Do you have a more specific question? This is a bit too broad.

**Reply by mgolokhov** [reply]
I'm missing a piece of the puzzle, how to pipe/connect BaseTableWidget, BaseTableProvider and CustomTableProvider, CusomTableProvider.

```dart
class CustomTable extends HookConsumerWidget {
  @override
  Widget build(BuildContext context, WidgetRef ref) {
    return BaseTableScreen(
      // we have multiple customTableProviders with different loading strategy and isolated state, still based on BasedTableProvider
      target: customTableProvider, 
    );
  }
}
```
What's the right choice of generics inside BaseTableScreen?
`StateNotifierProvider<?> target` , so we can use `ref.watch(target).when(...)`

**Reply by rrousselGit** [reply]
You're not giving me enough information here. What exactly are you trying to do?

**Reply by mgolokhov** [reply]
Generally, I would like to create BaseTableWidget with common functionality (load, refresh, sort, etc)
Also, I want a common state to decouple in BaseTableProvider.

Through an app, one reuses that BaseTableWidget by creating CustomTables with different loading strategies but with the same flow and states (loading, refreshing, error).

**Reply by mgolokhov** [reply]
Most examples show how to work with the specific provider as `final variable`  like `ref.watch(specificProvider).when(...handle states as union...)`
Is it possible to watch providers in a more dynamic way something like `ref.watch(baseProvider).when(...)`?

**Reply by mgolokhov** [reply]
For now, the only way I see that BaseTableWidget should not have BaseTableProvider inside and expose only callbacks
and a developer should make a bit more piping
ex `onBuild`
```dart
 child = ref.watch(customTableProvider).when(
          loading: () => buildLoadingView(),
          data: (TableViewModel newVm) => buildTableView(newVm),
          error: () => buildErrorView(),
          refreshing: (TableViewModel oldVm) => buildRefreshing(oldVm),
          );
```

Example for action `onRefresh`:
```dart
        ref.read(customTableProvider.notifier).refresh();
```

**Reply by rrousselGit** [reply]
I still don't quite understand what you're trying to do.

Are you trying to use Riverpod to avoid having to make statefulwidgets?

**Reply by mgolokhov** [reply]
Yes, it's one of the points.
How can I design a common Widget with Riverpod as state manager,
so a developer can reuse that Widget (reuse actions and state and some flow) with minimal boilerplate and still have the possibility to expand functionality.

**Reply by rrousselGit** [reply]
Riverpod isn't meant to be a solution for local widget state. Providers are for shared state only

If you want a local state framework, you can have a look at flutter_hooks, which is typically combined with Riverpod, but for local widget state.

**Comment by neiljaywarner** [comment]
i also would like to have some guidance here, for example is it "unwise" to put it in a shared components repo..
the tip I got from SO or something is making a provider static under the things it's providing
